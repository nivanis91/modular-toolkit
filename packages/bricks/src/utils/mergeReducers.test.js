import { createStore } from 'redux';
import { mergeReducers, createFakeActionAndReducer } from '.';
import combineReducers from '../combineReducers';

const [foo, fooAction] = createFakeActionAndReducer('FOO');
const [bar, barAction] = createFakeActionAndReducer('BAR');
const [baz, bazAction] = createFakeActionAndReducer('BAZ');
const [bazong, bazongAction] = createFakeActionAndReducer('BAZONG');

describe('When my original reducer is a simple function', () => {
    let originalReducer;
    beforeEach(() => (originalReducer = foo));
    describe('and I create a store', () => {
        let store;
        beforeEach(() => (store = createStore(originalReducer)));
        describe('the initial state', () => it('is correct', () => expect(store.getState()).toMatchSnapshot()));
        describe('and my additional reducers map contains just one reducer', () => {
            let additionalReducers;
            beforeEach(() => (additionalReducers = { bar }));
            describe('and I merge original and additional reducers', () => {
                let mergedReducer;
                beforeEach(() => (mergedReducer = mergeReducers(originalReducer, additionalReducers)));
                describe("and I replace the store's reducer with the merged reducer", () => {
                    beforeEach(() => store.replaceReducer(mergedReducer));
                    describe('the new state', () => it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    describe('and I dispatch an action for the original reducer', () => {
                        beforeEach(() => store.dispatch(fooAction));
                        describe('the resulting state', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                    describe('and I dispatch an action for the additional reducer', () => {
                        beforeEach(() => store.dispatch(barAction));
                        describe('the result', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                });
            });
        });
        describe('and my additional reducers map contains two reducers', () => {
            let additionalReducers;
            beforeEach(() => (additionalReducers = { bar, baz }));
            describe('and I merge original and additional reducers', () => {
                let mergedReducer;
                beforeEach(() => (mergedReducer = mergeReducers(originalReducer, additionalReducers)));
                describe("and I replace the store's reducer with the merged reducer", () => {
                    beforeEach(() => store.replaceReducer(mergedReducer));
                    describe('the new state', () => it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    describe('and I dispatch an action for the original reducer', () => {
                        beforeEach(() => store.dispatch(fooAction));
                        describe('the resulting state', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                    describe('and I dispatch an action for the first additional reducer', () => {
                        beforeEach(() => store.dispatch(barAction));
                        describe('the result', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                    describe('and I dispatch an action for the second additional reducer', () => {
                        beforeEach(() => store.dispatch(bazAction));
                        describe('the result', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                });
            });
        });
        describe('and my additional reducers map contains two nested reducers', () => {
            let additionalReducers;
            beforeEach(() => (additionalReducers = { bar, corge: { baz } }));
            describe('and I merge original and additional reducers', () => {
                let mergedReducer;
                beforeEach(() => (mergedReducer = mergeReducers(originalReducer, additionalReducers)));
                describe("and I replace the store's reducer with the merged reducer", () => {
                    beforeEach(() => store.replaceReducer(mergedReducer));
                    describe('the new state', () => it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    describe('and I dispatch an action for the original reducer', () => {
                        beforeEach(() => store.dispatch(fooAction));
                        describe('the resulting state', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                    describe('and I dispatch an action for the top additional reducer', () => {
                        beforeEach(() => store.dispatch(barAction));
                        describe('the result', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                    describe('and I dispatch an action for the nested additional reducer', () => {
                        beforeEach(() => store.dispatch(bazAction));
                        describe('the result', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                });
            });
        });
    });
});

describe('When my original reducer is a combined reducer', () => {
    let originalReducer;
    beforeEach(() => (originalReducer = combineReducers({ foo, bazong })));
    describe('and I create a store', () => {
        let store;
        beforeEach(() => (store = createStore(originalReducer)));
        describe('the initial state', () => it('is correct', () => expect(store.getState()).toMatchSnapshot()));
        describe('and my additional reducers map contains two nested reducers', () => {
            let additionalReducers;
            beforeEach(() => (additionalReducers = { bar, corge: { baz } }));
            describe('and I merge original and additional reducers', () => {
                let mergedReducer;
                beforeEach(() => (mergedReducer = mergeReducers(originalReducer, additionalReducers)));
                describe("and I replace the store's reducer with the merged reducer", () => {
                    beforeEach(() => store.replaceReducer(mergedReducer));
                    describe('the new state', () => it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    describe('and I dispatch an action for the first reducer of the combined original reducer', () => {
                        beforeEach(() => store.dispatch(fooAction));
                        describe('the resulting state', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                    describe('and I dispatch an action for the other reducer of the combined original reducer', () => {
                        beforeEach(() => store.dispatch(bazongAction));
                        describe('the resulting state', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                    describe('and I dispatch an action for the top additional reducer', () => {
                        beforeEach(() => store.dispatch(barAction));
                        describe('the result', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                    describe('and I dispatch an action for the nested additional reducer', () => {
                        beforeEach(() => store.dispatch(bazAction));
                        describe('the result', () =>
                            it('is correct', () => expect(store.getState()).toMatchSnapshot()));
                    });
                });
            });
        });
    });
});
