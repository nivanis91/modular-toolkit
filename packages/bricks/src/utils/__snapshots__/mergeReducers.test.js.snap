// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When my original reducer is a combined reducer and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the first reducer of the combined original reducer the resulting state is correct 1`] = `
Object {
  "bar": Object {},
  "bazong": Object {},
  "corge": Object {
    "baz": Object {},
  },
  "foo": Object {
    "value": "foo",
  },
}
`;

exports[`When my original reducer is a combined reducer and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the nested additional reducer the result is correct 1`] = `
Object {
  "bar": Object {},
  "bazong": Object {},
  "corge": Object {
    "baz": Object {
      "value": "baz",
    },
  },
  "foo": Object {},
}
`;

exports[`When my original reducer is a combined reducer and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the other reducer of the combined original reducer the resulting state is correct 1`] = `
Object {
  "bar": Object {},
  "bazong": Object {
    "value": "bazong",
  },
  "corge": Object {
    "baz": Object {},
  },
  "foo": Object {},
}
`;

exports[`When my original reducer is a combined reducer and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the top additional reducer the result is correct 1`] = `
Object {
  "bar": Object {
    "value": "bar",
  },
  "bazong": Object {},
  "corge": Object {
    "baz": Object {},
  },
  "foo": Object {},
}
`;

exports[`When my original reducer is a combined reducer and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer the new state is correct 1`] = `
Object {
  "bar": Object {},
  "bazong": Object {},
  "corge": Object {
    "baz": Object {},
  },
  "foo": Object {},
}
`;

exports[`When my original reducer is a combined reducer and I create a store the initial state is correct 1`] = `
Object {
  "bazong": Object {},
  "foo": Object {},
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains just one reducer and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the additional reducer the result is correct 1`] = `
Object {
  "bar": Object {
    "value": "bar",
  },
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains just one reducer and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the original reducer the resulting state is correct 1`] = `
Object {
  "bar": Object {},
  "value": "foo",
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains just one reducer and I merge original and additional reducers and I replace the store's reducer with the merged reducer the new state is correct 1`] = `
Object {
  "bar": Object {},
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the nested additional reducer the result is correct 1`] = `
Object {
  "bar": Object {},
  "corge": Object {
    "baz": Object {
      "value": "baz",
    },
  },
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the original reducer the resulting state is correct 1`] = `
Object {
  "bar": Object {},
  "corge": Object {
    "baz": Object {},
  },
  "value": "foo",
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the top additional reducer the result is correct 1`] = `
Object {
  "bar": Object {
    "value": "bar",
  },
  "corge": Object {
    "baz": Object {},
  },
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains two nested reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer the new state is correct 1`] = `
Object {
  "bar": Object {},
  "corge": Object {
    "baz": Object {},
  },
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains two reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the first additional reducer the result is correct 1`] = `
Object {
  "bar": Object {
    "value": "bar",
  },
  "baz": Object {},
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains two reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the original reducer the resulting state is correct 1`] = `
Object {
  "bar": Object {},
  "baz": Object {},
  "value": "foo",
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains two reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer and I dispatch an action for the second additional reducer the result is correct 1`] = `
Object {
  "bar": Object {},
  "baz": Object {
    "value": "baz",
  },
}
`;

exports[`When my original reducer is a simple function and I create a store and my additional reducers map contains two reducers and I merge original and additional reducers and I replace the store's reducer with the merged reducer the new state is correct 1`] = `
Object {
  "bar": Object {},
  "baz": Object {},
}
`;

exports[`When my original reducer is a simple function and I create a store the initial state is correct 1`] = `Object {}`;
